File: nni/runtime/tuner_command_channel/semantic_command.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

from dataclasses import dataclass
import dataclasses
import re
from typing import Any, Dict, Literal, Optional
import typeguard
import nni
from nni.runtime.tuner_command_channel.command_type import CommandType
from nni.typehint import _ParameterSearchSpace, TrialRecord

new_to_old ={'Initialize': CommandType.Initialize,
'RequestTrialJobs': CommandType.RequestTrialJobs,
'ReportMetricData': CommandType.ReportMetricData,
'UpdateSearchSpace': CommandType.UpdateSearchSpace,
'ImportData': CommandType.ImportData,
'AddCustomizedTrialJob': CommandType.AddCustomizedTrialJob,
'TrialEnd': CommandType.TrialEnd,
'Terminate': CommandType.Terminate,
'Ping': CommandType.Ping,
'Initialized': CommandType.Initialized,
'NewTrialJob': CommandType.NewTrialJob,
'SendTrialJobParameter': CommandType.SendTrialJobParameter,
'NoMoreTrialJobs': CommandType.NoMoreTrialJobs,
'KillTrialJob': CommandType.KillTrialJob
}

@dataclass
class BaseCommand:
    command_type: str

    def dump(self) -> str:
        command_dict = {key:val for key, val in self.__dict__.items() if key != 'command_type'}
        return nni.dump(command_dict)

    def validate(self):
        validate_type(self)

    def _to_legacy_command_type(self) -> tuple[CommandType, str]:
        if self.command_type not in new_to_old:
            raise AssertionError('Unsupported command type: {}'.format(self.command_type))
        old_command_type = new_to_old[self.command_type]

        command_dict = {key:val for key, val in self.__dict__.items() if key != 'command_type'}
        command_json = nni.dump(command_dict)
        return old_command_type, command_json

    @classmethod
    def load(cls, command_type, command_json: str):
        command_dict = {}
        if len(command_json) == 0:
            command_dict['command_type'] = command_type
            return cls(**command_dict)
        command_dict = nni.load(command_json)
        command_dict['command_type'] = command_type
        return cls(**command_dict)

@dataclass
class Initialize(BaseCommand):
    search_space: Dict[str, _ParameterSearchSpace]

    def _to_legacy_command_type(self) -> tuple[CommandType, str]:
        old_command_type = new_to_old[self.command_type]
        command_json = nni.dump(self.search_space)
        return old_command_type, command_json

    @classmethod
    def load(cls, command_type, command_json: str):
        command_dict = {}
        command_dict['search_space'] = nni.load(command_json)
        command_dict['command_type'] = command_type
        return cls(**command_dict)

@dataclass
class RequestTrialJobs(BaseCommand):
    job_num: int

    def _to_legacy_command_type(self) -> tuple[CommandType, str]:
        old_command_type = new_to_old[self.command_type]
        command_json = repr(self.job_num)
        return old_command_type, command_json

    @classmethod
    def load(cls, command_type, command_json: str):
        command_dict = {}
        command_dict['job_num'] = nni.load(command_json)
        command_dict['command_type'] = command_type
        return cls(**command_dict)

@dataclass
class ReportMetricData(BaseCommand):
    trial_job_id: str
    parameter_id: int
    sequence: int
    type: Literal['FINAL', 'PERIODICAL', 'REQUEST_PARAMETER']
    value: str
    parameter_index: Optional[int] = None
    parameters: Optional[Any] = None


@dataclass
class UpdateSearchSpace(BaseCommand):
    search_space: dict

@dataclass
class ImportData(BaseCommand):
    data:  list[TrialRecord]

@dataclass
class AddCustomizedTrialJob(BaseCommand):
    def _to_legacy_command_type(self) -> tuple[CommandType, str]:
        old_command_type = new_to_old[self.command_type]
        return old_command_type, ''

@dataclass
class TrialEnd(BaseCommand):
    trial_job_id: str
    event: str
    hyper_params: dict

    @classmethod
    def load(cls, command_type, command_json: str):
        hyper_params_json = command_json.split("hyper_params\":")[1][1:-2]
        command_dict = {}
        command_dict['hyper_params'] = nni.load(eval('\''+hyper_params_json+'\''))
        subcommand = re.findall(r':"(.*?)"', command_json.split("hyper_params\":")[0])
        command_dict['trial_job_id'] = subcommand[0]
        command_dict['event'] = subcommand[1]
        command_dict['command_type'] = command_type
        return cls(**command_dict)

@dataclass
class Terminate(BaseCommand):
    def _to_legacy_command_type(self) -> tuple[CommandType, str]:
        old_command_type = new_to_old[self.command_type]
        return old_command_type, ''

@dataclass
class Ping(BaseCommand):
    def _to_legacy_command_type(self) -> tuple[CommandType, str]:
        old_command_type = new_to_old[self.command_type]
        return old_command_type, ''

@dataclass
class Initialized(BaseCommand):
    def _to_legacy_command_type(self) -> tuple[CommandType, str]:
        old_command_type = new_to_old[self.command_type]
        return old_command_type, ''

@dataclass
class NewTrialJob(BaseCommand):
    parameter_id: int
    parameter_source: str
    parameters: Any
    parameter_index: Optional[int] = None
    placement_constraint: Optional[dict] = None
    version_info: Optional[str] = None

@dataclass
class SendTrialJobParameter(BaseCommand):
    parameter_id: int
    parameter_source: str
    parameters: Any
    trial_job_id: Optional[str] = None
    parameter_index: Optional[int] = None

@dataclass
class NoMoreTrialJobs(BaseCommand):
    parameter_id: Optional[int] = None
    parameter_source: Optional[str] = None
    parameters: Optional[Any] = None
    parameter_index: Optional[int] = None

@dataclass
class KillTrialJob(BaseCommand):
    trial_job_id: str

    def _to_legacy_command_type(self) -> tuple[CommandType, str]:
        old_command_type = new_to_old[self.command_type]
        return old_command_type, self.trial_job_id

def validate_type(command):
    class_name = type(command).__name__
    for field in dataclasses.fields(command):
        value = getattr(command, field.name)
        #check existense
        if is_missing(value):
            raise ValueError(f'{class_name}: {field.name} is not set')
        if not is_instance(value, field.type):
            raise ValueError(f'{class_name}: type of {field.name} ({repr(value)}) is not {field.type}')

def is_missing(value):
    """
    Used to check whether a dataclass field has ever been assigned.

    If a field without default value has never been assigned, it will have a special value ``MISSING``.
    This function checks if the parameter is ``MISSING``.
    """
    # MISSING is not singleton and there is no official API to check it
    return isinstance(value, type(dataclasses.MISSING))

def is_instance(value, type_hint):
    try:
        typeguard.check_type('_', value, type_hint)
    except TypeError:
        return False
    return True

File: ts/nni_manager/core/semanticCommand.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { PlacementConstraint } from "common/trainingService";
import { assert } from "console";
import { IMPORT_DATA, INITIALIZE, INITIALIZED, KILL_TRIAL_JOB, PING, 
    REPORT_METRIC_DATA, REQUEST_TRIAL_JOBS, 
    TERMINATE, TRIAL_END, UPDATE_SEARCH_SPACE } from "./commands";

export abstract class BaseCommand {
    public abstract toLegacyCommand(): string;
    public abstract validate(): void;
}

export class Initialize implements BaseCommand {
    searchSpace?: string;

    constructor(searchSpace?: string) {
        this.searchSpace = searchSpace;
    }

    public toLegacyCommand(): string {
        return INITIALIZE + this.searchSpace;
    }

    public validate(): void {
        throw new Error("Method not implemented.");
    }
}

export class RequestTrialJobs implements BaseCommand {
    jobNum: string;

    constructor(jobNum: number) {
        this.jobNum = String(jobNum);
    }

    public toLegacyCommand(): string {
        return REQUEST_TRIAL_JOBS + this.jobNum;
    }

    public validate(): void {
        assert(typeof this.jobNum === 'string');
    }
}

export class ReportMetricData implements BaseCommand {
    metricData: string;

    constructor(metricData: string) {
        this.metricData = metricData;
    }

    public toLegacyCommand(): string {
        return REPORT_METRIC_DATA + this.metricData;
    }

    public validate(): void {
        assert(typeof this.metricData === 'string');
    }
}

export class UpdateSearchSpace implements BaseCommand {
    searchSpace: string;

    constructor(searchSpace: string) {
        this.searchSpace = searchSpace;
    }

    public toLegacyCommand(): string {
        return UPDATE_SEARCH_SPACE + this.searchSpace;
    }

    public validate(): void {
        assert(typeof this.searchSpace === 'string');
    }
}

export class ImportData implements BaseCommand {
    data: string;

    constructor(data: string) {
        this.data = data;
    }

    public toLegacyCommand(): string {
        return IMPORT_DATA + this.data;
    }

    public validate(): void {
        assert(typeof this.data === 'string');
    }
}

export class TrialEnd implements BaseCommand {
    trialJobId: string;
    event: string;
    hyperParams: string;

    constructor(trialJobId: string, event: string, hyperParams: string) {
        this.trialJobId = trialJobId;
        this.event = event;
        this.hyperParams = hyperParams;
    }

    public toLegacyCommand(): string {
        return TRIAL_END + JSON.stringify({
            trial_job_id: this.trialJobId, // eslint-disable-line @typescript-eslint/camelcase
            event: this.event, 
            hyper_params: this.hyperParams});// eslint-disable-line @typescript-eslint/camelcase
    }

    public validate(): void {
        assert(typeof this.trialJobId === 'string');
        assert(typeof this.event === 'string');
        assert(typeof this.hyperParams === 'string');
    }
}

export class Terminate implements BaseCommand {
    public toLegacyCommand(): string {
        return TERMINATE;
    }

    public validate(): void {
        throw new Error("Method not implemented.");
    }
}

export class Ping implements BaseCommand {
    public toLegacyCommand(): string {
        return PING;
    }

    public validate(): void {
        throw new Error("Method not implemented.");
    }
}

export class Initialized implements BaseCommand {
    public toLegacyCommand(): string {
        return INITIALIZED;
    }

    public validate(): void {
        throw new Error("Method not implemented.");
    }
}

export class NewTrialJob implements BaseCommand {
    parameterId: number;
    parameters: object;
    parameterSource: string;
    parameterIndex?: number;
    placementConstraint?: PlacementConstraint;
    versionInfo?: string;

    constructor(content: NewTrialJobContent) {
        this.parameterId = content.parameter_id;
        this.parameters = content.parameters;
        this.parameterSource = content.parameter_source;
        this.parameterIndex = content.parameter_index;
        this.placementConstraint = content.placement_constraint;
        this.versionInfo = content.version_info;  
    }

    public toLegacyCommand(): string {
        throw new Error("Method not implemented.");
    }

    public validate(): void {
        throw new Error("Method not implemented.");
    }
}

export class SendTrialJobParameter implements BaseCommand {
    parameterSource: string;
    parameters: object;
    parameterIndex: number;
    trialJobId: string;
    parameterId?: number;

    constructor(content: SendTrialJobParameterContent) {
        this.trialJobId = content.trial_job_id;
        this.parameterId = content.parameter_id;
        this.parameterSource = content.parameter_source;
        this.parameters = content.parameters;
        this.parameterIndex = content.parameter_index;
    }

    public toLegacyCommand(): string {
        throw new Error("Method not implemented.");
    }

    public validate(): void {
        assert(this.parameterIndex >= 0);
        assert(this.trialJobId !== undefined);
    }
}

export class NoMoreTrialJobs implements BaseCommand {
    parameterSource?: string;
    parameters?: object;
    parameterIndex?: number;
    parameterId?: number;

    constructor(content: NoMoreTrialJobsContent) {
        this.parameterSource = content.parameter_source;
        this.parameters = content.parameters;
        this.parameterIndex = content.parameter_index;
        this.parameterId = content.parameter_id;
    }

    public toLegacyCommand(): string {
        throw new Error("Method not implemented.");
    }

    public validate(): void {
        throw new Error("Method not implemented.");
    }
}

export class KillTrialJob implements BaseCommand {
    trialJobId?: string;

    constructor(trialJobId?: string) {
        this.trialJobId = trialJobId;
    }

    public toLegacyCommand(): string {
        return KILL_TRIAL_JOB + this.trialJobId;
    }

    public validate(): void {
        throw new Error("Method not implemented.");
    }
}

export interface NoMoreTrialJobsContent {
    readonly parameter_id?: number;
    readonly parameter_source?: string;
    readonly parameters?: object;
    readonly parameter_index?: number;
}

export interface SendTrialJobParameterContent {
    readonly parameter_id: number;
    readonly parameter_source: string;
    readonly parameters: object;
    readonly parameter_index: number;
    readonly trial_job_id: string;
}

export interface NewTrialJobContent {
    readonly parameter_id: number;
    readonly parameters: object;
    readonly parameter_source: string;
    readonly parameter_index?: number;
    readonly placement_constraint?: PlacementConstraint;
    readonly version_info?: string;
}